sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/AgnosticEncoder.scala

```diff
@@ -107,6 +108,51 @@ object AgnosticEncoders {
     override def dataType: DataType = schema
   }
 
+  case class TupleEncoder[T](
```

Comment:
@zhenlineo we should be able to use `ProductEncoder` for this.

Reply:
Done. Any special case we missed?